### ISSUE TITLE
bitmonerod with GCC 6.1.1 not syncing on armv7h: block does not have enough proof of work .... unexpected difficulty

### ISSUE THREAD

**Issue by radfish** [monero-project/monero#847]
v0.9.4.0-a837c9c built natively with GCC v6

Synced on an x86_64 VPS (compiled from master) up to block 1050787, transferred lmdb to armv7h, launched daemon. It is stuck at block 1050788, with this error repeating:

```
Attempting to get output pubkey by global index, but key does not exist
EXCEPTION: Attempting to get output pubkey by global index, but key does not exist

[P2P9]Block with id: <74724176de696cfa20ccbaa5f7c203084e1ff963d19f6cf7f2a32d277dff5cd2>
does not have enough proof of work: <34f45fad9dfa1e314088b5c58342aa59be05e5972bd9c2a0cd200c8ecd2d7d0b>
unexpected difficulty: 1656389584
```

**Comment by moneroexamples** [comment]
Newest changes to monero code require resyncing blockchain, at least for me. Your issue  might be because of that.

**Comment by radfish** [comment]
@moneroexamples I synced on x86_64 VPS from scratch.

**Comment by radfish** [comment]
@hyc suggested that crypto code on ARM might be broken (due to latest GCC 6 or some other reason). The tests fail on my armv7h board (and another ARM server), which corroborates that hypothesis:
```
Running tests...
Test project bitmonero-git/src/bitmonero
      Start  1: hash-target
 1/13 Test  #1: hash-target ......................   Passed    0.38 sec
      Start  2: coretests
 2/13 Test  #2: coretests ........................***Failed  25731.55 sec
      Start  3: crypto
 3/13 Test  #3: crypto ...........................   Passed   97.99 sec
      Start  4: unit_tests
 4/13 Test  #4: unit_tests .......................***Failed    2.78 sec
      Start  5: difficulty
 5/13 Test  #5: difficulty .......................   Passed    0.05 sec
      Start  6: hash-fast
 6/13 Test  #6: hash-fast ........................   Passed    0.08 sec
      Start  7: hash-slow
 7/13 Test  #7: hash-slow ........................***Failed    1.31 sec
      Start  8: hash-tree
 8/13 Test  #8: hash-tree ........................   Passed    0.01 sec
      Start  9: hash-extra-blake
 9/13 Test  #9: hash-extra-blake .................   Passed    0.03 sec
      Start 10: hash-extra-groestl
10/13 Test #10: hash-extra-groestl ...............   Passed    0.04 sec
      Start 11: hash-extra-jh
11/13 Test #11: hash-extra-jh ....................   Passed    0.11 sec
      Start 12: hash-extra-skein
12/13 Test #12: hash-extra-skein .................   Passed    0.03 sec
      Start 13: libwallet_api_tests
13/13 Test #13: libwallet_api_tests ..............   Passed   21.08 sec

77% tests passed, 3 tests failed out of 13

Total Test time (real) = 25855.53 sec

The following tests FAILED:
          2 - coretests (Failed)
          4 - unit_tests (Failed)
          7 - hash-slow (Failed)
Errors while running CTest
Makefile:104: recipe for target 'test' failed
make: *** [test] Error 8
```

**Comment by iamsmooth** [comment]
comment out the ARM asm code for mul128 in slow-hash.c and try again. I have seen it fail on other gcc versions as well.

**Comment by radfish** [comment]
@iamsmooth Thanks, that worked. Documented the exact patch in #850.
Some tests still fail, but the above-described problem is gone, node does seem to be syncing:

```
Running tests...
Test project /mnt/flext/dev/bitmonero-git/src/bitmonero/build
      Start  1: hash-target
 1/12 Test  #1: hash-target ......................   Passed    1.13 sec
      Start  2: crypto
 2/12 Test  #2: crypto ...........................   Passed  353.81 sec
      Start  3: unit_tests
 3/12 Test  #3: unit_tests .......................***Failed    2.23 sec
      Start  4: difficulty
 4/12 Test  #4: difficulty .......................   Passed    0.36 sec
      Start  5: hash-fast
 5/12 Test  #5: hash-fast ........................   Passed    0.16 sec
      Start  6: hash-slow
 6/12 Test  #6: hash-slow ........................   Passed    3.30 sec
      Start  7: hash-tree
 7/12 Test  #7: hash-tree ........................   Passed    0.02 sec
      Start  8: hash-extra-blake
 8/12 Test  #8: hash-extra-blake .................   Passed    0.04 sec
      Start  9: hash-extra-groestl
 9/12 Test  #9: hash-extra-groestl ...............   Passed    0.13 sec
      Start 10: hash-extra-jh
10/12 Test #10: hash-extra-jh ....................   Passed    0.20 sec
      Start 11: hash-extra-skein
11/12 Test #11: hash-extra-skein .................   Passed    0.04 sec
      Start 12: libwallet_api_tests
12/12 Test #12: libwallet_api_tests ..............   Passed   43.54 sec

92% tests passed, 1 tests failed out of 12

Total Test time (real) = 405.02 sec

The following tests FAILED:
          3 - unit_tests (Failed)
Errors while running CTest

```

**Comment by radfish** [comment]
For the sake of completeness:

I applied the patch. Then I synced on x86_64 from scratch (yes, again), cleanly exited bitmonerod, trasferred the bitmonerod folder to armv7h box, and launched bitmonerod. Shortly after startup, the log does contain the error `EXCEPTION: Attempting to get output pubkey by global index, but key does not exist`, but it does NOT contain the error about `block does not have enough proof of work`.

The node appears to be working fine: `status` shows the top block, and it gets new blocks from the network:
```
$ bitmonerod status
Height: 1056797/1056797 (100.0%) on mainnet, not mining, net hash 12.47 MH/s, v2, up to date, 6+4 connections
$ bitmonerod status
Height: 1056798/1056799 (100.0%) on mainnet, not mining, net hash 12.46 MH/s, v2, up to date, 7+4 connections
$ bitmonerod status
Height: 1056799/1056799 (100.0%) on mainnet, not mining, net hash 12.44 MH/s, v2, up to date, 7+5 connections
$ bitmonerod status
Height: 1056800/1056802 (100.0%) on mainnet, not mining, net hash 12.41 MH/s, v2, up to date, 3+13 connections
```

**Comment by radfish** [comment]
So far it seems that the `EXCEPTION: ...` is not fatal (although not nice), because the node goes on to sync fine and work fine.  But the `block does not have enough proof of work` error is fatal, the node keeps fetching the same batch of blocks and keeps rejecting it with this error in an endless loop, while saturating disk I/O.

So far we (@hyc) found that from the same commit we can build a broken binary (with both the nonfatal and the fatal error) and a working binary (with the nonfatal error only). The prime suspect difference seems to be GCCv6 vs GCC 4.8/4.9.

**Comment by radfish** [comment]
Tested with `-fno-strict-aliasing`: all tests pass (didn't try coretests, cause it's long), import proceeds past block #4 (see above), and syncing from network seems to proceed ok so far.

Will submit a PR with `-fno-strict-aliasing` as a temporary workaround, until aliasing is fixed in the code (leaving this ticket open for that).

**Comment by hyc** [comment]
This was finally fixed as issue #1991 in PR #2078

**Comment by moneromooo-monero** [comment]
+resolved
